### ISSUE TITLE
Names shadowed by transitive dependencies

### ISSUE THREAD

**Issue by SimonJF** [links-lang/links#606]
Just as I'd merged the last patch, I tried running my larger application (should have done that first -- will know for future), and ran into the following issue.

Essentially, `open` in one module affects name resolution in another module. As an example, consider the three following files:

`one.links`:

```
fun concat(xs) {
  fold_right((^^), "", xs)
}
```

`two.links`:
```
open One;

fun foo() {
  concat(["Hello", "World"])
}
```

`three.links`:
```
open Two;

fun main() {
  concat([[1], [2], []])
}

main()
```

In `one.links`, we define `concat` which shadows the `concat` from the prelude. It now has the type `([String]) ~> String` instead of `([[a]]) ~> [a]`.

Running `links three.links`, I now get:

```
[simon@dazzle mfe-shadowing]$ links three.links 
registering driver for null
compiling to IR
compiled IR
Finishing process MAIN
Unification error: Couldn't match ([String]) ~> String against ([[Int]]) -@ _::Any
three.links:4: Type error: The function
    `concat'
has type
    `([String]) ~a~> String'
while the arguments passed to it have types
    `[[Int]]'
and the currently allowed effects are
    `|b::Any'
In expression: concat([[1], [2], []]).
```
where I would expect `[1, 2]`.

**Comment by dhil** [comment]
Yes, it is subtle. Dynamic binding is flexible, but in its unrestricted form it is inherently anti-modular.

**Comment by SimonJF** [comment]
Right -- this seems counterintuitive.

Just to copy a relevant comment from #607:

"Currently, open subsumes import, but not vice versa. In a subsequent patch I will change the semantics of open such that it can no longer be used to require an external module."

Would that then mean that:

`one.links`:

```
fun concat(xs) {
  fold_right((^^), "", xs)
}
```

`two.links`:
```
import One;
open One;

fun foo() {
  concat(["Hello", "World"])
}
```

`three.links`:
```
import Two;

fun main() {
  concat([[1], [2], []])
}

main()
```

would be allowed and function the way I would expect it to? More specifically, would an `import` followed by an `open` of an external module work as expected?

**Comment by dhil** [comment]
> would be allowed and function the way I would expect it to? More specifically, would an import followed by an open of an external module work as expected?

An `import` followed by an `open` will cause the source file to be loaded twice, presently. I will change the behaviour, such that `open` cannot cause a file load to happen. In patch #607 I have added syntactic sugar, as suggested by @jamescheney in #571, for simultaneous import and opening of a compilation unit, i.e. `import open One;` both resolves `One` and pulls its contents into the current scope.

**Comment by dhil** [comment]
Just to clarify: the reason for separating the changes into two patches is that, the changing the semantics of `open` is a breaking change.

**Comment by SimonJF** [comment]
> An import followed by an open will cause the source file to be loaded twice, presently. I will change the behaviour, such that open cannot cause a file load to happen.
Sure. My question was was assuming if you did the aforementioned change to the semantics of `open`.

> the changing the semantics of open is a breaking change.
Yes -- I suppose my rationale in opening this issue was that I didn't expect the semantics of `open` to change in this way: before, this shadowing via transitive dependencies didn't happen.

Is #607 ready for review?

**Comment by dhil** [comment]
The semantics of `open` is essentially that of `include` from OCaml. Because the chaser/loader performs a whole program transformation when it resolves dependencies. If you read the transformed program from top to bottom, you will expect to see the behaviour that you are experiencing. But of course, as a programmer, you reason locally, and hence you think of the different source files as being separate, isolated entities. This is a limitation due to the current loader/chaser. 

The new loader actually process each source file separately, thus once I get that in, you should see the behaviour that you expect.

**Comment by SimonJF** [comment]
That makes sense. Thanks for the explanation.

In that case, let's keep this open to document this change in behaviour until the new loader is merged?

**Comment by dhil** [comment]
> Yes -- I suppose my rationale in opening this issue was that I didn't expect the semantics of open to change in this way: before, this shadowing via transitive dependencies didn't happen.

However strange it may sound, this strange semantics is in fact the "correct" semantics with respect to how we transform programs. As mentioned above, we inadvertently get "include" semantics because the chaser performs a whole-program transformation.

> In that case, let's keep this open to document this change in behaviour until the new loader is merged?

Yes, this is a loader problem, not a module problem.

**Comment by dhil** [comment]
I mistakenly concluded that this was a loader problem, but as pointed out by you, the chaser combines the main program and its dependencies by adding the dependencies as synthetic modules to the main problem, hence an `open` statement in the global scope of a dependency should not affect the scope of any other dependency nor the main program. 

Fixed by #609.
